### PR TITLE
Developed preference xml inspections

### PR DESCRIPTION
**Pull request by Iamwade**
**Description** (*)
Added  inspection to the di.xml file for preference.
The inspection is performed for preference attributes "for" and "type". 
If the class in the specified path for these attributes does not exist, then the corresponding warning will be displayed
image

image

**Fixed Issues (if relevant)**

1. Fixes magento/magento2-phpstorm-plugin#566

**Contribution checklist** (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with integration/functional tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by Iamwade** [comment]
Waiting for https://github.com/magento/magento2-phpstorm-plugin/pull/577

**Comment by Iamwade** [comment]
Waiting for #582

File: resources/inspectionDescriptions/PreferenceXmlInspections.html [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+<p>
+  Validates classes and interfaces for  attributes "for" and the "type" for the "preference" tag
+</p>
+<p>This inspection checks the "for" and "type" attribute of the &ltpreference/&gt; tag for:</p>
+<ul>
+  <li>the correct path to the class or interface</li>
+  <li>existence of attribute values</li>
+</ul>
+</body>
+</html>
```
Comment (bohdan-harniuk) [inline resources/inspectionDescriptions/PreferenceXmlInspections.html:18]: @Iamwade, please add an empty line in the and of the file.

File: resources/inspectionDescriptions/PreferenceXmlInspections.html [hunk 2]
```diff
@@ -0,0 +1,18 @@
+<!--
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+-->
+<html>
+<body>
+<p>
+  Validates classes and interfaces for  attributes "for" and the "type" for the "preference" tag
```
Comment (bohdan-harniuk) [inline resources/inspectionDescriptions/PreferenceXmlInspections.html:10]: @Iamwade, please remove extra space after the first "for". And if you can, please, formulate this sentence differently. For now, it is hard to read.

File: resources/META-INF/plugin.xml [hunk 3]
```diff
@@ -219,6 +219,13 @@
                          enabledByDefault="true" level="WARNING"
                          implementationClass="com.magento.idea.magento2plugin.inspections.xml.InvalidDependencyInjectionTypeInspection"/>
 
+        <localInspection language="XML" groupPath="XML"
+                         shortName="PreferenceXmlInspections"
+                         displayName="Inspection the existence of PHP classes for attributes in the preference"
```
Comment (bohdan-harniuk) [inline resources/META-INF/plugin.xml:224]: @Iamwade, please extract displayName attribute value to the magento2.inspection bundle and reference it from this context:
```suggestion
                         bundle="magento2.inspection" key="inspection.displayName.PreferenceXmlInspections"
```

File: resources/META-INF/plugin.xml [hunk 4]
```diff
@@ -219,6 +219,13 @@
...
+                         groupName="Magento 2"
```
Comment (bohdan-harniuk) [inline resources/META-INF/plugin.xml:225]:
```suggestion
                         groupBundle="magento2.inspection" groupKey="inspection.group.name"
```

File: src/com/magento/idea/magento2plugin/inspections/xml/PreferenceDeclarationInspection.java [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+
+@SuppressWarnings({"PMD.ExcessiveMethodLength", "PMD.NPathComplexity"})
+public class PreferenceDeclarationInspection extends XmlSuppressableInspectionTool {
+
+    @Override
+    public @NotNull PsiElementVisitor buildVisitor(
+            final @NotNull ProblemsHolder problemsHolder,
+            final boolean isOnTheFly
+    ) {
+        return new XmlElementVisitor() {
```
Comment (bohdan-harniuk) [inline src/com/magento/idea/magento2plugin/inspections/xml/PreferenceDeclarationInspection.java:32]:
```suggestion
        return new XmlElementVisitor() {
        
```

File: src/com/magento/idea/magento2plugin/inspections/xml/PreferenceDeclarationInspection.java [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+            private final InspectionBundle inspectionBundle = new InspectionBundle();
+
```
Comment (bohdan-harniuk) [inline src/com/magento/idea/magento2plugin/inspections/xml/PreferenceDeclarationInspection.java:34]:
```suggestion
```

File: tests/com/magento/idea/magento2plugin/inspections/xml/PreferenceDeclarationInspectionTest.java [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+
+package com.magento.idea.magento2plugin.inspections.xml;
+
+import com.magento.idea.magento2plugin.magento.files.ModuleDiXml;
+
+public class PreferenceDeclarationInspectionTest extends InspectionXmlFixtureTestCase {
+
+    private static final String ARGUMENT_VALUE_IS_EMPTY =
+            "inspection.error.idAttributeCanNotBeEmpty";
+
```
Comment (bohdan-harniuk) [inline tests/com/magento/idea/magento2plugin/inspections/xml/PreferenceDeclarationInspectionTest.java:14]: @Iamwade, do we really need extra empty lines after each constant property?

File: tests/com/magento/idea/magento2plugin/inspections/xml/PreferenceDeclarationInspectionTest.java [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+    public void setUp() throws Exception {
+        super.setUp();
+        myFixture.enableInspections(PreferenceDeclarationInspection.class);
+    }
+
+    /**
+     * Test for an error for the "type" attribute because it is empty.
+     * <preference for="" type="Foo\Bar\Model\Logger"/>
+     */
+    public void testAttrArgForValuesIsEmpty() {
```
Comment (bohdan-harniuk) [inline tests/com/magento/idea/magento2plugin/inspections/xml/PreferenceDeclarationInspectionTest.java:37]: @Iamwade, sorry, but I cannot understand this...you are testing type attribute, but method annotation, description, name and test data file name says that you are testing For attribute. Could you check it?

File: tests/com/magento/idea/magento2plugin/inspections/xml/PreferenceDeclarationInspectionTest.java [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+        );
+
+        assertHasHighlighting(forAttrIsEmptyMessage);
+    }
+
+    /**
+     * Test for an error for the "for" attribute because it is empty.
+     * <preference for="Foo\Bar\Model\Logger" type=""/>
+     */
+    public void testAttrArgTypeValueIsEmpty() {
```
Comment (bohdan-harniuk) [inline tests/com/magento/idea/magento2plugin/inspections/xml/PreferenceDeclarationInspectionTest.java:52]: @Iamwade, there is the same confusing...you are testing for attribute, but the method, test data directory and XML in the description opposite. Please, check it.

File: tests/com/magento/idea/magento2plugin/inspections/xml/PreferenceDeclarationInspectionTest.java [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+        );
+
+        assertHasHighlighting(forAttrIsEmptyMessage);
+    }
+
+    /**
+     * Test for an no error for the "for" attribute because this class exists.
+     * <preference for="Foo\Bar\Model\Logger" type=""/>
+     */
+    public void testAttrForClassExists() {
```
Comment (bohdan-harniuk) [inline tests/com/magento/idea/magento2plugin/inspections/xml/PreferenceDeclarationInspectionTest.java:67]: @Iamwade, check it, please. (the reason the same as for previous methods)

File: tests/com/magento/idea/magento2plugin/inspections/xml/PreferenceDeclarationInspectionTest.java [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+    /**
+     * Test for an no error for the "type" attribute because this class exists.
+     * <preference for="" type="Foo\Bar\Model\Logger"/>
+     */
+    public void testAttrTypeClassExists() {
+        configureFixture();
+
+        final String typeAttrIsEmptyMessage =  inspectionBundle.message(
+                ARGUMENT_VALUE_IS_EMPTY,
+                ModuleDiXml.PREFERENCE_ATTR_FOR
```
Comment (bohdan-harniuk) [inline tests/com/magento/idea/magento2plugin/inspections/xml/PreferenceDeclarationInspectionTest.java:87]: @Iamwade, check it, please. (the reason the same as for previous methods)

**Review by bohdan-harniuk — CHANGES_REQUESTED** [review]
Hello, @Iamwade! I've left few comments in your PR. Please, check them.

File: resources/inspectionDescriptions/PreferenceXmlInspections.html (hunk 1 above)
Comment (Iamwade) [reply resources/inspectionDescriptions/PreferenceXmlInspections.html]: Fixed.

File: resources/inspectionDescriptions/PreferenceXmlInspections.html (hunk 2 above)
Comment (Iamwade) [reply resources/inspectionDescriptions/PreferenceXmlInspections.html]: Fixed.

File: resources/META-INF/plugin.xml (hunk 3 above)
Comment (Iamwade) [reply resources/META-INF/plugin.xml]: Fixed.

File: resources/META-INF/plugin.xml (hunk 4 above)
Comment (Iamwade) [reply resources/META-INF/plugin.xml]: Fixed.

File: src/com/magento/idea/magento2plugin/inspections/xml/PreferenceDeclarationInspection.java (hunk 5 above)
Comment (Iamwade) [reply src/com/magento/idea/magento2plugin/inspections/xml/PreferenceDeclarationInspection.java:32]: Fixed.

File: tests/com/magento/idea/magento2plugin/inspections/xml/PreferenceDeclarationInspectionTest.java (hunk 8 above)
Comment (Iamwade) [reply tests/com/magento/idea/magento2plugin/inspections/xml/PreferenceDeclarationInspectionTest.java]: Fixed.

File: tests/com/magento/idea/magento2plugin/inspections/xml/PreferenceDeclarationInspectionTest.java (hunk 9 above)
Comment (Iamwade) [reply tests/com/magento/idea/magento2plugin/inspections/xml/PreferenceDeclarationInspectionTest.java:48]: Fixed.

File: tests/com/magento/idea/magento2plugin/inspections/xml/PreferenceDeclarationInspectionTest.java (hunk 11 above)
Comment (Iamwade) [reply tests/com/magento/idea/magento2plugin/inspections/xml/PreferenceDeclarationInspectionTest.java]: Fixed.

File: resources/inspectionDescriptions/PreferenceXmlInspections.html (hunk 1 above)
Comment (bohdan-harniuk) [reply resources/inspectionDescriptions/PreferenceXmlInspections.html]: Thank you!

File: resources/inspectionDescriptions/PreferenceXmlInspections.html (hunk 2 above)
Comment (bohdan-harniuk) [reply resources/inspectionDescriptions/PreferenceXmlInspections.html]: Thank you!

File: resources/META-INF/plugin.xml (hunk 3 above)
Comment (bohdan-harniuk) [reply resources/META-INF/plugin.xml]: Thank you!

File: resources/META-INF/plugin.xml (hunk 4 above)
Comment (bohdan-harniuk) [reply resources/META-INF/plugin.xml]: Thank you!

File: src/com/magento/idea/magento2plugin/inspections/xml/PreferenceDeclarationInspection.java (hunk 5 above)
Comment (bohdan-harniuk) [reply src/com/magento/idea/magento2plugin/inspections/xml/PreferenceDeclarationInspection.java:32]: Thank you!

File: tests/com/magento/idea/magento2plugin/inspections/xml/PreferenceDeclarationInspectionTest.java (hunk 7 above)
Comment (bohdan-harniuk) [reply tests/com/magento/idea/magento2plugin/inspections/xml/PreferenceDeclarationInspectionTest.java]: Thank you!

File: tests/com/magento/idea/magento2plugin/inspections/xml/PreferenceDeclarationInspectionTest.java (hunk 8 above)
Comment (bohdan-harniuk) [reply tests/com/magento/idea/magento2plugin/inspections/xml/PreferenceDeclarationInspectionTest.java]: Thank you!

File: tests/com/magento/idea/magento2plugin/inspections/xml/PreferenceDeclarationInspectionTest.java (hunk 10 above)
Comment (bohdan-harniuk) [reply tests/com/magento/idea/magento2plugin/inspections/xml/PreferenceDeclarationInspectionTest.java:63]: Thank you!

File: tests/com/magento/idea/magento2plugin/inspections/xml/PreferenceDeclarationInspectionTest.java (hunk 9 above)
Comment (bohdan-harniuk) [reply tests/com/magento/idea/magento2plugin/inspections/xml/PreferenceDeclarationInspectionTest.java:48]: Thank you!

File: tests/com/magento/idea/magento2plugin/inspections/xml/PreferenceDeclarationInspectionTest.java (hunk 11 above)
Comment (bohdan-harniuk) [reply tests/com/magento/idea/magento2plugin/inspections/xml/PreferenceDeclarationInspectionTest.java]: Thank you!

File: testData/inspections/xml/PreferenceDeclarationInspection/classAttrForDoesNotExists/di.xml [hunk 12]
```diff
@@ -0,0 +1,5 @@
+<?xml version="1.0"?>
+<config xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+        xsi:noNamespaceSchemaLocation="urn:magento:framework:ObjectManager/etc/config.xsd">
+    <preference for="" type="Not\Existent\Class"/>
```
Comment (bohdan-harniuk) [inline testData/inspections/xml/PreferenceDeclarationInspection/classAttrForDoesNotExists/di.xml:4]:
```suggestion
    <preference for="Not\Existent\Class" type=""/>
```

**Review by bohdan-harniuk — CHANGES_REQUESTED** [review]
Hello, @Iamwade! I've left one more comment in your PR. Please, check it.

File: testData/inspections/xml/PreferenceDeclarationInspection/classAttrForDoesNotExists/di.xml (hunk 12 above)
Comment (Iamwade) [reply testData/inspections/xml/PreferenceDeclarationInspection/classAttrForDoesNotExists/di.xml]: @bohdan-harniuk fixed. Thank you.

**Review by bohdan-harniuk — DISMISSED** [review]
@Iamwade, thank you for your work here!

**Review by VitaliyBoyko — CHANGES_REQUESTED** [review]
Getting an error when removing the attribute value

Screenshot 2021-05-04 at 15 06 40

**Comment by Iamwade** [comment]
@VitaliyBoyko a bug when removing an attribute value has been fixed. In the process of fixing this error, I found this error for the <type> tag in the file 
https://github.com/magento/magento2-phpstorm-plugin/blob/4.0.0-develop/src/com/magento/idea/magento2plugin/inspections/xml/InvalidDependencyInjectionTypeInspection.java
 and I fixed it too.
 Please review.

**Review by VitaliyBoyko — APPROVED** [review]

**Pull request merged**